torch/csrc/dynamo/eval_frame.c

```diff
@@ -461,57 +461,55 @@ inline static PyObject* eval_custom_code(
   DEBUG_NULL_CHECK(tstate);
   DEBUG_NULL_CHECK(frame);
   DEBUG_NULL_CHECK(code);
+  DEBUG_CHECK(nlocals_new >= nlocals_old);
+
+  PyObject* result = NULL;
+
   #if IS_PYTHON_3_11_PLUS
+
   DEBUG_CHECK(ncells == frame->f_code->co_ncellvars);
   DEBUG_CHECK(nfrees == frame->f_code->co_nfreevars);
-  #else
-  DEBUG_CHECK(ncells == PyTuple_GET_SIZE(frame->f_code->co_cellvars));
-  DEBUG_CHECK(nfrees == PyTuple_GET_SIZE(frame->f_code->co_freevars));
-  #endif
-  DEBUG_CHECK(nlocals_new >= nlocals_old);
 
-  PyFrameObject* shadow_obj = PyFrame_New(tstate, code, frame->f_globals, NULL);
-  #if IS_PYTHON_3_11_PLUS
-  THP_EVAL_API_FRAME_OBJECT* shadow = shadow_obj->f_frame;
-  Py_XINCREF(frame->f_func->func_closure);
-  shadow->f_func->func_closure = frame->f_func->func_closure;
-  // NOTE: in Python 3.11.1+, PyFrame_New changes prev_instr, causing
-  // Python runtime errors, so we revert the change.
-  // See https://github.com/python/cpython/pull/97886.
-  // TODO if there are more shadow frame related bugs in the future,
-  // consider not using PyFrame_New.
-  shadow->prev_instr = _PyCode_CODE(shadow->f_code) - 1;
-  #else
-  THP_EVAL_API_FRAME_OBJECT* shadow = shadow_obj;
-  #endif
+  PyFunctionObject *func = _PyFunction_CopyWithNewCode((PyFunctionObject *) frame->f_func, code);
+  if (func == NULL) {
+    return NULL;
+  }
+
+  size_t size = code->co_nlocalsplus + code->co_stacksize + FRAME_SPECIALS_SIZE;
+  THP_EVAL_API_FRAME_OBJECT* shadow = malloc(size);
   if (shadow == NULL) {
-    Py_DECREF(shadow_obj);
+    Py_DECREF(func);
     return NULL;
   }
 
-  #if IS_PYTHON_3_11_PLUS
+  Py_INCREF(func);
+  // consumes reference to func
+  _PyFrame_InitializeSpecials(shadow, func, NULL, code->co_nlocalsplus);
+
   PyObject** fastlocals_old = frame->localsplus;
   PyObject** fastlocals_new = shadow->localsplus;
 
-  // copy from old fastlocals to new fastlocals:
+  for (int i = 0; i < code->co_nlocalsplus; i++) {
+    fastlocals_new[i] = NULL;
```

Comment:
Why is this needed? Is that in case you don't re-populate the whole array you don't want it to be uninitialized?

Reply:
Yes. Every element in the localsplus array will be Py_XINCREF'd.

Reply:
Can you have a comment here explaining that please?